src/domain/runner/store.rs

```diff
@@ -0,0 +1,372 @@
+use super::constants::{LOAD_COLD, LOAD_WARM, STORE_BASE, STORE_NEW, STORE_REFUND, STORE_UPDATE};
+use wasmer::RuntimeError;
+
+pub const STORAGE_POINTER_SIZE: usize = 32;
+pub const STORAGE_VALUE_SIZE: usize = 32;
+pub type StoragePointer = [u8; STORAGE_POINTER_SIZE];
+pub type StorageValue = [u8; STORAGE_VALUE_SIZE];
+pub const STORAGE_VALUE_ZERO: [u8; STORAGE_VALUE_SIZE] = [0; STORAGE_VALUE_SIZE];
+
+pub struct CacheResponse {
+    pub value: StorageValue,
+    pub gas_cost: u64,
+    pub gas_refund: i64,
+}
+
+#[derive(Debug, Clone)]
+pub struct CacheValue {
+    original: StorageValue,
+    current: StorageValue,
+}
+
+impl CacheValue {
+    // New value filled from existing store
+    pub fn new(value: StorageValue) -> Self {
+        Self {
+            original: value,
+            current: value,
+        }
+    }
+}
+
+impl CacheResponse {
+    pub fn new(value: [u8; STORAGE_VALUE_SIZE], gas_cost: u64, gas_refund: i64) -> Self {
+        Self {
+            value,
+            gas_cost,
+            gas_refund,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Cache {
+    values: std::collections::HashMap<[u8; STORAGE_POINTER_SIZE], CacheValue>,
+    reads: usize,
+    writes: usize,
+}
+
+impl Cache {
+    pub fn new() -> Self {
+        Self {
+            values: std::collections::HashMap::new(),
+            reads: 0,
+            writes: 0,
+        }
+    }
+
+    #[allow(dead_code)]
+    pub fn exists(&self, pointer: &[u8; STORAGE_POINTER_SIZE]) -> bool {
+        self.values.contains_key(pointer)
+    }
+
+    pub fn get<GetFun>(
+        &mut self,
+        pointer: &[u8; STORAGE_POINTER_SIZE],
+        get_fn: GetFun,
+    ) -> Result<CacheResponse, RuntimeError>
+    where
+        GetFun: Fn(StoragePointer) -> Result<StorageValue, RuntimeError>,
+    {
+        self.reads += 1;
+
+        // If object exists in the cache
+        if let Some(value) = self.values.get(pointer) {
+            Ok(CacheResponse::new(value.current, LOAD_WARM, 0))
+        }
+        // first acces of pointer
+        else {
+            let original_value = get_fn(*pointer)?;
+            self.values
+                .insert(*pointer, CacheValue::new(original_value));
+
+            Ok(CacheResponse::new(original_value, LOAD_COLD, 0))
+        }
+    }
+
+    pub fn set<GetFun, SetFun>(
+        &mut self,
+        key: [u8; STORAGE_POINTER_SIZE],
+        value: [u8; STORAGE_VALUE_SIZE],
+        get_value: GetFun,
+        set_value: SetFun,
+    ) -> Result<CacheResponse, RuntimeError>
+    where
+        GetFun: Fn(StoragePointer) -> Result<StorageValue, RuntimeError>,
+        SetFun: Fn(StoragePointer, StorageValue) -> Result<(), RuntimeError>,
+    {
+        // increase number of writes
+        self.writes += 1;
```

Comment:
Ethereum has a minimum of gas for this operation (I am not sure, if we want to have same behavior due to gas refund calculation)